BK2SIT~1.6/BK2SIT~1.6/BK2SITE.1
.\" Copyright (C) 1999 Jose M. Vidal
.\" 
.\" Permission is granted to any individual or institution to use, copy, or
.\" redistribute this software so long as all of the original files are
.\" included, that it is not sold for profit, and that this copyright notice
.\" is retained.
.\" http://bk2site.sourceforge.net
.\" bk2site-dev@lists.sourceforge.net
.\" groff -man -Tascii bk2site.1
.\"
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.ta .3i .6i .9i 1.2i 1.5i 1.8i
.TH bk2site 1 "19 January 2000"
.SH NAME 
bk2site \- convert a Netscape Navigator bookmark file into a yahoo-like website with slashdot-like news.
.SH SYNOPSIS
.B bk2site [-nd] [-f 
.I config-file
.B ] [-d
.I dest-dir
.B ] [-t
.I name
.B ] [-f1
.I basetemplate
.B ] [-f1
.I othertemplate
.B ] [-o
.I outputfile
.B ] 
.B ] [-xbel
.I outputfile
.B ][-ns] [-nc] [-old
.I days
.B ] [--help] [
.I mainbookmarkfile
.B ] [
.I bookmarkfile
.B ...]
.SH DESCRIPTION
Converts a Netscape Navigator bookmark file into a yahoo-like website
with slashdot-like news. It is typically run as a cron job so as to
keep your website automatically updated. Customization is achieved
thru the template files and the .bk2siterc file, as explained
below. All the options can also be given in the .bk2siterc file.
.SH OPTIONS
.IP -n
Don't use a subdirectory hierachy. Use ugly long filenames. This
option is useful for tree depth limited providers like geocities.
.IP "-f config-file"
Use the configuration file
.I config-file
instead of
.IR ~/.bk2siterc
and
.IR /etc/bk2site/dot.bk2siterc .
.IP "-d dest-dir"
Local directory for generated files.  Default is /home/httpd/html/ or
as given in
.IR /etc/bk2site/dot.bk2siterc
and 
.IR ~/.bk2siterc 
.IP "-t name" 
Name of top-level bookmark folder to use. If ommited bk2site will
convert whole bookmark file.
.IP "-f1 basetemplate" 
Index template file for top index.  
.IP "-f2 othertemplate"
Template for the other pages. 
.IP "-o outputfile"
Output bookmark file. All the bookmarks that form the website are
output to this file in Netscape bookmark file format.
.IP "-xbel outputfile"
Output bookmark file in XML Bookmarks Exchange Language (XBEL).
All the bookmarks that form the website are output to this file.
.IP -ns
Do NOT generate the website. This is useful if you just want to
generate the bookmarkfile or some other files, as specified in
~/.bk2siterc.
.IP -nc
Do NOT put HTML comments in the output.
.IP "-old days"
Number of days (default is 30) after which a bookmark is no longer
new.

.SH FILES
.I /etc/bk2site/dot.bk2siterc
.RS
The system configuration file. It is always read first.
.RE
.I ~/.bk2siterc
.RS
The user configuration file. It is read after the global configuration 
file and before the one specified in the -f argument. Values set later 
always override previously set values. That is, the user configuration 
file overrides values set in the globla configuration file.

.SH Quick Start
The minimum 
.I ~/.bk2siterc 
you will need, if you run your own webserver with files in
/home/http/html (as is the default in most distributions) is:
.EX
destdir /home/httpd/html
search /home/httpd/cgi-bin/urls.db
extrafilebase /etc/bk2site/newbase.html
extrafilename new.html
.EE

.SH SUMMARY
bk2site simply reads your bookmarks file and generates a bunch of
files which form a yahoo-like site, with the option of using one of
your folders to generate a set of slashdot-like news.

In order to customize its output, you need to set up (at least)
three files.
.I indexbase.html
.I otherbase.html
and your own 
.I ~/.bk2siterc.
Samples of these three files are included with the distribution and
should be found under /etc/bk2site or /usr/doc/bk2site.

.IP indexbase.html 
This file forms the template for your top-level
index. That is, the page everyone will see first. This file must be a
legal html file and contain the comments as described below.

.IP otherbase.html
It is the template for all the other files. It also needs to have
comments, as described below.

.IP ~/.bk2siterc.
Edit the 
.I /etc/bk2site/dot.bk2siterc 
file and copy it to
.I ~/.bk2siterc
The sample file has lots of comments, so read it. Also,
a sample
.I ~/.bk2siterc
is given at the end of this manual page. This file tells
bk2site where to find indexbase.html and otherbase.html
.PP
There can also be extra configuration files if you wish to generate
extra output (for example, you might wish to generate an xml file to
be read by MyNetscape.com, or a plain text file to be used in an
automate monthly newsletter). The extra template files are given with
the "extrafilebase" and "extrafilename" variables in your
.I ~/.bk2siterc

Basically, bk2site supports 5 different sections in your document. 
Each section refers to the type of URLS that will be output by bk2site 
at that point. They are:
.IP folder
folders from your bookmarkfile
.IP url
urls (leafs) from your bookmarkfile
.IP newadd
urls of the most recently, sorted by most-recently-created
.IP news
urls from the newstopfolder, sorted by most-recently-created.
.IP tophits
urls with the most hits, sorted by number of hits (as given by urllog)
.IP newsandnewadd
combines the urls in news and newadd and sorts them by most-recently-created.
.PP
Each section is replaced by bk2site with the appropiate urls. For
example, the folder section is replaced with the URLs for the
appropiate folders, while the url section is replaced by the leaf
urls. Exactly how these URLs will look depends on the values you give
to the different variables within each section.

The easiest way to understand how it works is to look at at the sample
files. Lets look at the folders (in Yahoo! these are known as the
"categories"). In the sample otherbase.html you will find something
like:
.EX
<!--bk2site:begin folder-->
<!--bk2site:top <table width="100%"><tr><td><ul>-->
<!--bk2site:template <li><A HREF="%URL"><B>%TITLE</B></A> <I>(%LEAFS)</I> %CONDDASH <small>%COMMENT</small>-->
<!--bk2site:separator </ul></td><td><ul>-->
<!--bk2site:bottom </ul></td></tr></table><hr>-->
<!--bk2site:cols 2-->
<!--bk2site:maxnum 0-->
<!--bk2site:end-->
.EE

These comments guide how bk2site generates the html for the
folders. Each section must always start with the <!--bk2site:begin
section-type--> and end with <!--bk2site:end-->. Within each section
section-type--> there are a number of variables (e.g. top, template,
section-type--> separator), each one of these variables is followed by 
section-type--> the value we are assigning to it.

The whole section will be replaced, in the output file, by the code
generated by bk2site. The section begins with: 

.IP "<!--bk2site:begin folder-->"
Tells bk2site that this is the place where we want to place the
folders.
.PP
After the begin statemend we see a number of variables, each one is
assigned a value. The function of each variable is:

.IP top 
If there are folders to print, print the value of this variable (in
this case <tablewidth="100%"><tr><td><ul>) before any folders.

.IP template
Each folder is printed using the template given by the value of this
variable (in this case <li><A
HREF="%URL"><B>%TITLE</B></A><I>(%LEAFS)</I> %CONDDASH
<small>%COMMENT</small>). You can insert many directives, which are
all in capitals and preceded with a %. These directives will be
replaced with the appropiate values. All the directives are explained
below.

.IP cols
Tells us to print the folders in colums. In this example we are
setting this value to 2. This simply inserts the value of "separator"
after the appropiate number of elements.

.IP separator
After each column is over, we print this separator in order to start
the next colum.

.IP bottom 
If there are folders to print, print the value of this
variable after all folders have been printed.

.IP maxnum
The maximum number of folders to print. If 0 or ommited, then we print 
all the folder that exist in the bookmark file.

.IP daysold
Do not include any folders or urls that were created more than
15 days ago. If 0 or omitted then we print all folders.

.IP between
Places the contents of this variable between every two items, but not
at the beginning or at the end; and not when the separator is being
printed. This is sometimes useful.

.IP startnum
This would tell bk2site to start printing on the 15th element.
.PP
There are other special comments that can appear anywhere in a
template file and do not need to be within a section. They will be
replaced. In their place bk2site will put some appropiate HTML code.

.IP "<!--bk2site:navigatebar-->"
Gets replaced by the navigation
bar. This is only useful in otherbase.html. This is not a variable and 
should not be set to any value. 

.IP "<!--bk2site:title-->"
Gets replaced by the title (current folder
name), prepended by the "title" defined in .bk2siterc.

.IP "<!--bk2site:include /home/me/myfile.html-->"
Includes myfile.html at
the place where the comment is placed. myfile.html can itself include
any of the tags. It eliminates any occurances of %QUERY, %ESCQUERY and
%NUMBER in the included file. Its a great way to include headers and
%footers that are common to all your pages. Also, instead of a file
you can also specify a URL.

.IP "<!--bk2site:include:search /home/me/myfile.html-->" 
As above, but
leaves %QUERY, %ESCQUERY and %NUMBER. Also, instead of a file
you can also specify a URL.

.IP "<!--bk2site:date-->"
Gets replaced by the current date and time.

.IP "<!--bk2site:path-->"
Gets replaced by the path to the current file being written. This
is only useful in otherbase base pages since the path we are talking
about is just the path that bk2site creates to mirror your folder structure.
That is, it is NOT the absolute path on your filesystem, it is only
the set of directories that must be traversed from the top of the
tree in order to get to the current page. This comment is useful when
you need to reference the page you are on. (e.g. asking google to find
other pages related to that one.)

.PP
You can get sample copies of base files, showing how all these ugly
comment can be used, at the bk2site homepage
http://bk2site.sourceforge.net

.SH Directives
These are replaced for their appropiate values when they appear in a
template. They get instantiated for each url or folder.

.IP %URL
the URL
.IP %NOACCENTURL
the URL where all accented, and other strange characters, have been
replaced by equivalent values.
.IP %TITLE
the title
.IP %NOACCENTTITLE
the title where all accented, and other strange characters, have been
replaced by equivalent values.
.IP %COMMENT
the comment,
.IP %FILLCOMMENT
the comment, its justified for 80 columns, to make it look nice.
.IP %NOACCENTCOMMENT
the comment where all accented, and other strange characters, have been
replaced by equivalent values.
.IP %SHORTCOMMENT(x)
the first x characters of the comment, where x is an integer.
.IP %LONGCOMMENT
the comment in one long line with any carriage-returns or newlines
removed.
.IP %NHCOMMENT 
the comment but with everything between < and > stripped out
that is, all the HTML stripped out (No HTML Comment). Also on 80 columns.
.IP %FSCOMMENT
the first sentence of the comment, that is, everything up to the
first period. A single letter with a period will be skipped. If there is
no period then the whole comment.
.IP %JSCOMMENT
the comment but with the appropiate characters escaped so that it can
be used as a JavaScript string.
.IP %JSFSCOMMENT
the first sentence of the comment, that is, everything up to the
first period. A single letter with a period will be skipped. If there is
no period then the whole comment. Special characters are escaped so that
the comment can be used as a JavaScript string.
.IP %DAYCRE
the day created
.IP %DAYVIS
the day last visited
.IP %DAYMODM
the day last Modified
.IP %MONTHCRE1
the month created, in number
.IP %MONTHCRE2
the month created, in English
.IP  %MONTHVIS1
the month visited, in number
.IP %MONTHVIS2
the month visited, in English
.IP %MONTHMOD1
the month modified, in number
.IP %MONTHMOD2
the month modified, in English
.IP %YEARCRE
last two digits of the year created
.IP %YEARFCRE
the year created, including all digits (the full number).
.IP %YEARVIS
last two digits of the year visited
.IP %YEARFVIS
the year visited, including all digits (the full number).
.IP %YEARFMOD
last two digits of the year modified
.IP %YEARMOD
the year modified, including all digits (the full number).
.IP %LEAFS
the total number of leafs under it (for folders only)
.IP %NEW
the new.gif, if the time is right.
.IP %TIMEFCRE
the time created in full Wed Jun 30 21:49:03 1993
.IP %TIMEFMOD
the time modified in full Wed Jun 30 21:49:03 1993
.IP %TIMEFVIS
the time visited in full Wed Jun 30 21:49:03 1993
.IP %CONDDASH
prints a dash "-" if the comment is not null
.IP %IFCOMHAS(string)(html)
if the comment has "string" then insert "html" at point.
also eliminate "string" from comment.
e.g. %IFCOMHAS(*coolsite*)(<img src="/images/cool.gif">)   or
%IFCOMHAS(-neat-)(<font color="red">Neat Site</font>)
If you use it, you must use it for ALL templates
.IP %HITS
the number of hits this url has receives, as gathered from urllogfile
.PP
There are also functions you can use. Right now, these functions
are NOT available for search.pl so you should not use them in
your searchurltemplate variable in your .bk2siterc. 

Each function takes one or more arguments and gets replaced by the
results of applying that function to the given arguments. The function are:

.IP &FILL[arg|cols]
where arg is a string and cols is an integer. This function auto-fills
the given argument to cols columns. e.g. &FILL[%COMMENT].
.IP &NOHTML[arg]
where arg is a string. Gets rid of all the HTML in the argument. That is,
everything between < and >.
.IP &REPLACE[arg|from|to]
where arg, from, and to are strings. It replaces all instances of 
"from" with "to" in "arg".
.IP &CUT[arg|chars]
where arg is a string and chars is an integer. It returns the first
"chars" characters in "arg".
.IP &NOLINEBREAKS[arg]
where arg is a string. It replaces all newlines, returns and tabs
with spaces, in arg.
.IP &NOACCENTS[arg]
where arg is a string. It replaces all accented characters with their
non-accented counterparts. It also gets rid of other troublesome
characters.
.IP &FS[arg]
where arg is a string. It returns the first sentence in arg.
.PP

These functions can be nested, as in &CUT[&NOHTML[%COMMENT]|500]. The
strings used as arguments to these functions cannot include the characters
[, ], or | since that will confuse the parser.

.SH Netscape Navigator Comments
As you know, navigator allows you to give comments to bookmars and
folders. We use these to guide the behavior of bk2site (see below and
read dot.bk2siterc, esp %IFCOMHAS) and we also print them.

You can also use HTML in the comment and it will work, but:

*********WARNING********************WARNING***********

DO NOT PUT "HREF=" IN YOUR COMMENT. There is a bug in Navigator that
will make it mungle your bookmarks if you do that. The bug has been
reported. 

If you must, use "HREF =". The space is VERY important. I actually
like to use "href =". The use of lowercase and a space makes me feel
doubly safe.

*********WARNING********************WARNING***********

.SH Using Aliases
bk2site supports url aliases, as used by navigator. 

It also implements directory aliases. Since Navigator does not allow
us to create an alias for a folder/directory, we must do it with a
little hack.

Say you have a folder called "people" and you want to put an alias to
it somewhere else. You should go to where you want the alias to appear 
and create an empty folder called "people", then add the comment ALIAS
in the comment box for that folder.

That's it, the empty folder "people" will now appear with an @ at the
end of it (in the website) and will point to the real "people"
folder. You can have as many ALIAS folders as you want, pointing to
the same or different folders.

There's also support for directory aliases where the ALIAS name is
different from the real folder. So if you have a folder called
"robots" as a subfolder of your "Business" folder and you want to put an
alias to it called "Commercial" into your "AI" / "robotics" folder,
go to "robotics", create an empty folder called "Commercial", and finally
add the comment "ALIAS robots" in the comment box for that folder.

.SH Setting Private Folders
If there is a particular entry or folder that you want bk2site to
ignore simply give it the comment PRIVATE. The comment must be all
in capital letters and must be the only text on the comment field.

.SH Sorting Bookmarks
You can also have bk2site automatically sort the bookmarks in a
folder. This is done by adding the appropiate comments to the
folder. The valid values are:

.IP "sort title normal"
sorts based on title, alphabetically.
.IP "sort title inverse"
in inverse alphabetical on title.
.IP "sort creation normal"
Sort based on the time when the bookmark was
created. The more recent ones come first.
.IP "sort creation inverse"
more recent come last.
.IP "sort modified normal"
Sort based on the time when the bookmark was
last modified. The more recent ones come
first.
.IP "sort modified inverse"
more recent come last.
.IP "sort visit normal"
Sort based on the time when the bookmark was
last visited. The more recent ones come
first.
.IP "sort visit inverse"
more recent come last.
.IP "sort hits normal"
Sort based on the number of hits, as determined
by reading the urllog file. The ones with more hits
come first
.IP "sort hits inverse"
Less hits come first.
.PP

These comments are case-insensitive.

Remember that the folders only have a title and a creation date, so
sorting them on visit or modified will not do anything to them.

.SH Output of Bookmark file

You can use the -o switch, or the "outputbookmarkfile" variable in
.bk2siterc, to tell bk2site to output a bookmark file that contains
only the bookmarks that are shown on the website. That is, it excludes
those that were not under the main folder and those that where under
PRIVATE folders. For example
.EX
bk2site -o /tmp/published-bookmarks.html
.EE
will write out to /tmp/published-bookmarks.html. This is useful if you
want to give someone else a bookmark file containing only your public

BK2SIT~1.6/BK2SIT~1.6/BK2SITE.C
// bk2site by Jose M. Vidal
// http://jmvidal.ece.sc.edu  vidal@sc.edu
//
// Converts the Netscape bookmark.html file into a yahoo-like
// web site.
//
// Copyright 1999 Jose M. Vidal
// Jose M. Vidal, vidal@multiagent.com, http://jmvidal.ece.sc.edu
// This program is free software.  You can redistribute it and/or modify
// it under the terms of the GNU General Public License
//
// CONTRIBUTORS:
//  See the AUTHORS file
//
//
// $Id: bk2site.C,v 1.49 2000/12/15 23:52:56 jmvidal Exp $
// -*- mode:C++; tab-width:4; c-basic-offset:2; indent-tabs-mode:nil -*- 

//TO DO:
// - Escape accented charactes in name2filename (name2url)
// - Add ALIASID= and ALIASOF= options to directory comments, so as
//   to allow more flexible directory aliasing.
// - Add support for separators <HR>
// - urllog should have "time url IP" and bk2site should count only one hit
///   per IP.
//
// BUGS
// - Make search.pl ignore html marking....how????

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <fstream>
#include <iomanip>
#include <cstdlib>
#include <string>

// #if TIME_WITH_SYS_TIME
// # include <sys/time.h>
// # include <time.h>
// #else
// # if HAVE_SYS_TIME_H
// #  include <sys/time.h>
// # else
// #  include <time.h>
// # endif
// #endif 
#include <ctime>

#include "channel.H" //this include must appear before #include "reference.H"
#include "reference.H"


//defined in reference.C
extern string trim(const string & s);
extern int cmpNoCase(const string& s, const string &s2);
extern int replaceAllNC(string & s, string s1, string s2);
extern int replaceAll(string & s, string s1, string s2);

const string version = VERSION;
using namespace std;

void printUsage(const string & s)
{
  cout << "Usage: " << s << ": [OPTIONS] [main-bookmark-file] [bookmarkfile]*" << endl 
       << "\t If main-bookmark-file is omitted we use ~/.netscape/bookmarks.html" << endl
       << "\t Any extra bookmark files given are merged with main, see the README" << endl << endl;
  cout << "Builds a yahoo-like website based on one or more Navigator bookmark files" << endl << endl
       << "-nd\t\t - don't use a subdirectory hierachy. Use ugly long filenames" << endl
       << "\t\t  it's useful for tree depth limited providers like geocities" << endl
    //	 << "-bf <book file>\t - File with bookmarks (~/.netscape/bookmarks.html)" << endl
       << "-f <config file> - Name of configuration file (~/.bk2siterc)" << endl
       << "-d <dest dir>\t - Local directory for generated files." << endl
       << "\t\t   Default is /home/httpd/html/" << endl
       << "-t <name>\t - Name of top-level bookmark folder to use." << endl
       << "\t\t   If ommited will convert whole bookmark file." << endl
       << "-f1 <file>\t - Index template file for top index. " << endl
       << "-f2 <file>\t - Template for the other pages. " << endl 
       << "-o <file>\t - Output bookmark file." << endl
			 << "-xbel <file>\t - Output an xbel file." << endl 
       << "-ns\t\t - Do NOT generate the website. " << endl 
       << "-nc\t\t - Do NOT put HTML comments in the output." << endl
       << "-old <int>\t - Number of days (30) after which a bookmark is no longer new." << endl << endl
       << "Version " << version << endl << endl;
  exit(1);
}

void readConfigStream(istream & configStream, string varValues[],
											vector<string> & extraFileBase, vector<string> & extraFileName,
											channelContainer & channels)
{
	string curVarName;
	string curVarValue;

	string channelName ="";
	string channelUrl;
	string channelFile;
	while (configStream >> curVarName) {
		getline(configStream, curVarValue);
		curVarValue = trim(curVarValue);
		if (curVarName[0] == '#') continue;
		if (curVarValue[0] == '"') //get rid of surrounding "'s
	    curVarValue.erase(0,1);
		if (curVarValue[curVarValue.size() -1] == '"')
	    curVarValue.erase(curVarValue.size() - 1,1);

		if (getenv("HOME") != NULL)
			replaceAll(curVarValue,"$HOME", getenv("HOME"));
		else
			replaceAll(curVarValue,"$HOME", "");

		if (getenv("USER") != NULL)
			replaceAll(curVarValue,"$USER", getenv("USER"));
		else
			replaceAll(curVarValue,"$USER", "");

		if (getenv("LOGNAME") != NULL)
			replaceAll(curVarValue,"$LOGNAME", getenv("LOGNAME"));
		else
			replaceAll(curVarValue,"$LOGNAME", "");

		int i, found = 0;
		for (i = 0; i < NUMVARS; i++){
	    if (cmpNoCase(varNames[i],curVarName) == 0){
				varValues[i] = curVarValue;
				//	cout << varNames[i] << ":" << varValues[i] << endl;
				found = 1;
				break;
	    }
		};
		//these two are special, since they can appear many times.
		if (cmpNoCase("EXTRAFILEBASE",curVarName) == 0){
			extraFileBase.push_back(curVarValue);
			found = 1;
		}
		else if (cmpNoCase("EXTRAFILENAME",curVarName) == 0){
			extraFileName.push_back(curVarValue);
			found = 1;
		};

		//these can also appear many times
		if (cmpNoCase("CHANNELNAME",curVarName) == 0){
			if (channelName != ""){ //push back old one
				channels.addNewChannel(channelName, channelUrl, channelFile);
			}
			else
				channelName = curVarValue;
			found = 1;
		}
		else if (cmpNoCase("CHANNELURL",curVarName) == 0){
			channelUrl = curVarValue;
			found = 1;
		}
		else if (cmpNoCase("CHANNELFILE",curVarName) == 0){
			channelFile = curVarValue;
			channels.addNewChannel(channelName, channelUrl, channelFile);
			channelName = channelFile = channelUrl = "";
			found = 1;
		};
		if (found == 0)
	    cerr << "ERROR: variable " << curVarName << " is not a recognized name." << endl;
	}
	if (extraFileName.size() != extraFileBase.size()){
		cerr << "ERROR: the number of extrafilename declarations is not the same as the " << endl
				 << "       number of extrafilebase declarations. Exiting." << endl;
		exit(1);
	}
}

int main (int argc, char **argv)
{
	//initiallize default values.
	string varValues[NUMVARS];
	varValues[NEWGIF] = "<img src=\"/icons/new.gif\" alt=\"Added %MONTHCRE1/%DAYCRE/%YEARCRE\">";
	varValues[DESTDIR] = "";
	varValues[TOPFOLDER] = ""; 
	varValues[BOOKMARKFILE] = "bookmarks.html";
	varValues[INDEXFILENAME] = "indexbase.html";
	varValues[OTHERFILENAME] = "otherbase.html";
	varValues[TIMECUTOFF] = "30";
	varValues[TOP] = "Top";
	varValues[INDEX] = "index.html";
	varValues[SEARCH] = "";
	varValues[SEARCHTOROOTPATH] = "../";
	varValues[URLLOGFILE] = "";
	varValues[SUBDIRSEP] = "/";
	varValues[BACKDIRSEP] ="../";
	varValues[SEARCHURLTEMPLATE] = "<A HREF=\"%URL\">%TITLE</A> %NEW %CONDDASH %COMMENT";
	varValues[NEWSTOPFOLDER] = "";
	varValues[TITLE] = "My Bookmarks";
	varValues[OUTPUTBOOKMARKFILE] = "";
	varValues[COMMENTS] = "yes";
	varValues[FOLDERCREATION] = "maxdescendants";
	varValues[TOPHITSINCLUDENEWS] = "no";

	string xbelFileName = "";

	string homeDir = getenv("HOME");

	vector<string> extraFileBase;
	vector<string> extraFileName;

	vector<string> bookmarkfiles;
	int numBookmarkfiles = 0;
	int generateSite = 1;

    //read /etc/bk2site/dot.bk2siterc
    //    string globalConfigFile("CONFIGFILE");
    //    cout << "Checking " << "CONFIGFILE" << endl; //the -D on the compiler does not work. (egcs?)
	string globalConfigFile(CONFIGFILE);
	// cout << "Checking " << CONFIGFILE << endl;
	//the -D on the compiler does not work. (egcs?)
	// string globalConfigFile("/etc/bk2site/dot.bk2siterc");
	//string globalConfigFile("/etc/bk2site/dot.bk2siterc");
	ifstream globalConfigStream(globalConfigFile.c_str(), ios::in);
	channelContainer channels;
	if (globalConfigStream) 
		readConfigStream(globalConfigStream, varValues, extraFileBase, extraFileName, channels);

	int readHomeConfig = 1;
	for (int i = 1; i < argc; i++){
		string s(argv[i]);
		if (s == "-f")
			readHomeConfig = 0;
	}

	//read ~/.bk2siterc
	string configFile = homeDir + "/.bk2siterc";
	if (readHomeConfig == 1) {
		cout << "Reading " << configFile << endl;
		ifstream configStream(configFile.c_str(), ios::in);
		if (configStream) 
			readConfigStream(configStream, varValues, extraFileBase, extraFileName, channels);
	};

	//read the file  in -f
	for (int i = 1; i < argc; i++){
		string s(argv[i]);
		if (s == "-f")
	    if (++i < argc){
				configFile = argv[i];
				cout << "Reading " << configFile << endl;
				ifstream configStream2(configFile.c_str(), ios::in);
				if (configStream2)
					readConfigStream(configStream2, varValues, extraFileBase, extraFileName, channels);
				else {
					cerr << "ERROR: Could not open " << configFile << endl;
					return 1;
				};
	    }
	}

	//now read the command line arguments.
	for (int i = 1; i < argc; i++){
		string s(argv[i]);
		if (s == "-nd") {
	    varValues[SUBDIRSEP] = "_";
	    varValues[BACKDIRSEP] ="";
		}
		else if (s == "-ns"){
			generateSite = 0;
		}
		else if (s == "-nc"){
			varValues[COMMENTS] = "no";
		}
		else if (s == "-d")
	    if (++i < argc)
				varValues[DESTDIR] = argv[i];
	    else 
				printUsage(argv[0]);
		else if (s == "-t")
	    if (++i < argc)
				varValues[TOPFOLDER] = argv[i];
	    else
				printUsage(argv[0]);
		else if (s == "-f1")
	    if (++i < argc)
				varValues[INDEXFILENAME] = argv[i];
	    else
				printUsage(argv[0]);
		else if (s == "-f2")
	    if (++i < argc)
				varValues[OTHERFILENAME] = argv[i];
	    else
				printUsage(argv[0]);
		else if (s == "-old")
	    if (++i < argc){
				varValues[TIMECUTOFF] = argv[i];
	    }
	    else
				printUsage(argv[0]);
		else if (s == "-o")
	    if (++i < argc){
	      varValues[OUTPUTBOOKMARKFILE] = argv[i];
	    }
	    else
				printUsage(argv[0]);
		else if (s == "-xbel")
	    if (++i < argc){
				xbelFileName = argv[i];
	    }
	    else
				printUsage(argv[0]);
		else if (s == "-f") {
	    ++i;
	    continue;
		}
		else if (s == "-bf") {
	    if (++i < argc)
				varValues[BOOKMARKFILE] = argv[i];
	    else
				printUsage(argv[0]);
		}
		else if ((s == "-h") || (s == "--help") || (s == "-help") || (s == "-v")) {
			printUsage(argv[0]);
		}
		else {
			if (numBookmarkfiles == 0){
				varValues[BOOKMARKFILE] = s;
			}
			else
				bookmarkfiles.push_back(s);
			numBookmarkfiles++;
		}
	}
	int last;
	if (varValues[DESTDIR] == "")  
		varValues[DESTDIR] = "/home/httpd/html/";
	last = varValues[DESTDIR].length() -1;
	if (varValues[DESTDIR][last] != '/') varValues[DESTDIR] = varValues[DESTDIR] + "/";
	if (varValues[BOOKMARKFILE] == "")
		varValues[BOOKMARKFILE] = homeDir + "/.netscape/bookmarks.html";
	last =varValues[SEARCHTOROOTPATH].length() -1;
	if (varValues[SEARCHTOROOTPATH][last] != '/') 
		varValues[SEARCHTOROOTPATH] =  varValues[SEARCHTOROOTPATH] + "/";
	referenceTree rt;
    
	//    for (int i = 0; i <= NUMVARS; i++)
	//	cout << varNames[i] << "\t" << varValues[i] << endl;
	if (varValues[BOOKMARKFILE] != "") {
		iwebstream istream(varValues[BOOKMARKFILE].c_str(), ios::in);
		if (istream == 0){
			cout << "Could not open " << varValues[BOOKMARKFILE] << endl;
			return 1;
		}
		cout << "Reading " << varValues[BOOKMARKFILE] << endl;
		istream >> rt;
		cout << "Total Bookmarks in file= " << rt.getNumLeafs() << endl;
	}

	//    cout << rt;
	referenceTree * myrt = rt.getSubtree(varValues[TOPFOLDER]);
	if (myrt == 0) {
		cerr << "ERROR: " << varValues[TOPFOLDER] << " not found in " << varValues[BOOKMARKFILE] << endl;
		return 1;
	}


	//ugly hack to append the top folder title to the tree.
	// we must undo this hack after all the merges.
	referenceTree myrtTemp;
	reference refTemp;
	if (varValues[TOPFOLDER] != ""){
		refTemp.title = varValues[TOPFOLDER];
		refTemp.children = myrt;
		refTemp.folder = true;
		refTemp.aliasOf = "";
		refTemp.priv = false;
		myrtTemp.addReference(refTemp); //adds a copy of refTemp?, myrtTemp is a copy of myrt with new root.
		myrt = &myrtTemp;
		myrt->fixNumChildren();
	}

	//Merge all the other trees in
	for (vector<string>::const_iterator fn = bookmarkfiles.begin(); fn != bookmarkfiles.end(); fn++){
		referenceTree ort;
		const string & bfname = * fn;
		iwebstream obf(bfname.c_str(), ios::in);
		if (obf == 0){
			cout << "Could not open " << bfname << endl;
			return 1;
		}
		cout << "Reading " << bfname << endl;
		obf >> ort;
		cout << "Total Bookmarks in file = " << ort.getNumLeafs()<< endl;
		cout << "Merging " << bfname << " into main tree." << endl;
		myrt->merge(ort);
		cout << "Total Bookmarks in merged file = " << myrt->getNumLeafs() << endl;
	} 

	//undo the ugly hack
	if (varValues[TOPFOLDER] != ""){
		myrt = myrt->getFirstGrandchild();
		refTemp.children = 0;
	}

	//read the urllog file if needed
	if (varValues[URLLOGFILE] != ""){ 
		iwebstream istream(varValues[URLLOGFILE].c_str(), ios::in);
		vector<string> theURLS;
		vector<time_t> theTimes;
		if (istream == 0){ //If file exists but is empty
			cout << "Could not open " << varValues[URLLOGFILE] << " for reading." << endl;
		}
		else {
			cout << "Reading " << varValues[URLLOGFILE] << endl;
			time_t curTime;
			time_t cutoffTime = time(0) - atoi(varValues[HITSTIMECUTOFF].c_str());
			string currentURL = "";
			while (!(istream.eof())){
				istream >> curTime;
				istream >> currentURL;
				//	    cout << curTime << "-" << currentURL << endl;
				//       if (('0' <= currentURL[0] ) && ('9' >= currentURL[0])) { //starts with a num, then its not a URL
				// 	int numhits = atoi(currentURL.c_str());
				// 	myrt->increaseHits(oldURL, numhits - 1); //we already added 1
				//       }
				//       else
				// 	myrt->increaseHits(currentURL, 1);
				//       oldURL = currentURL;

				//if url starts with a number (its a time) or url is null
				//  keep trying to read a time/url pair.
				while (((currentURL[0] >= '0' && currentURL[0] <= '9') || currentURL == "") && !istream.eof()){
					curTime = atoi(currentURL.c_str());
					istream >> currentURL;
				}
	    
				if (curTime >= cutoffTime){
					channels.increaseHits(currentURL, 1);
					myrt->increaseHits(currentURL, 1);
					theURLS.push_back(currentURL);
					theTimes.push_back(curTime);
				}
			}
			//write out only the ones after cutoffTime;
			istream.close();
		}
		ofstream ostream(varValues[URLLOGFILE].c_str(), ios::out);
		cout << "Opening " << varValues[URLLOGFILE] << " for writing" << endl;
		if (ostream == 0){
	    cout << "Could not open " << varValues[URLLOGFILE] << " for writing" << endl;
	    return 1;
		}
		for (unsigned int i=0; i < theURLS.size(); i++)
	    ostream << theTimes[i] << "\t" << theURLS[i] << endl;
		ostream.close();
	}

	//Fix Aliases.
	myrt->fixAliases(myrt, "", varValues, "");

	//Set the creationtime for the folders to be the max of creation time of descendants
	// or children.
	if (varValues[FOLDERCREATION] == "maxdescendants")
		myrt->setFolderCreationToMaxDescendant();
	else if (varValues[FOLDERCREATION] == "maxchildren")
		myrt->setFolderCreationToMaxChildren();

	if (generateSite){
		myrt->createSite(varValues, channels);

		for (unsigned int ne = 0; ne < extraFileName.size(); ne++) 
			myrt->createPage(extraFileBase[ne], varValues[DESTDIR] + extraFileName[ne], varValues, channels);
	}

	//write out the bookmark file    
	if (varValues[OUTPUTBOOKMARKFILE] != "") {
		ofstream otstream(varValues[OUTPUTBOOKMARKFILE].c_str(), ios::out);
		cout << "Opening " << varValues[OUTPUTBOOKMARKFILE] << " for writing" << endl;
		if (otstream == 0){
			cout << "Could not open " << varValues[OUTPUTBOOKMARKFILE] << " for writing" << endl;
			return 1;
		}
		myrt->writeAsBookmarkFile(otstream, varValues[TITLE]);
	}


	//write out the xbel file    
	if (xbelFileName != "") {
		ofstream otstream(xbelFileName.c_str(), ios::out);
		cout << "Opening " << xbelFileName << " for writing" << endl;
		if (otstream == 0){
			cout << "Could not open " << xbelFileName << " for writing" << endl;
			return 1;
		}
		myrt->writeAsXBELFile(otstream, varValues[TITLE]);
	}


	return 0;
}
